lib/srv/alpnproxy/azure_msi_middleware.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package alpnproxy
+
+import (
+	"context"
+	"crypto"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/jwt"
+)
+
+// AzureMSIMiddleware implements a simplified version of MSI server serving auth tokens.
+type AzureMSIMiddleware struct {
+	// Identity is the Azure identity to be served by the server. Only single identity will be provided.
+	Identity string
+	// TenantID to be returned in a claim. Doesn't have to match actual TenantID as recognized by Azure.
+	TenantID string
+	// ClientID to be returned in a claim.
+	ClientID string
+
+	// Key used to sign JWT
+	Key crypto.Signer
+
+	// Clock is used to override time in tests.
+	Clock clockwork.Clock
+	// Log is the Logger.
+	Log logrus.FieldLogger
+}
+
+var _ LocalProxyHTTPMiddleware = &AzureMSIMiddleware{}
+
+func (m *AzureMSIMiddleware) OnStart(ctx context.Context, lp *LocalProxy) error {
+	if m.Log == nil {
+		m.Log = lp.cfg.Log
+	}
+
+	if m.Clock == nil {
+		m.Clock = lp.cfg.Clock
+	}
+
+	return m.CheckAndSetDefaults()
+}
+
+func (m *AzureMSIMiddleware) CheckAndSetDefaults() error {
+	if m.Key == nil {
+		return trace.BadParameter("missing Key")
+	}
+	if m.Identity == "" {
+		return trace.BadParameter("missing Identity")
+	}
+	if m.TenantID == "" {
+		return trace.BadParameter("missing TenantID")
+	}
+	if m.ClientID == "" {
+		return trace.BadParameter("missing ClientID")
+	}
+	return nil
+}
+
+func (m *AzureMSIMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
```

Comment:
Currently `AzureMSIMiddleware` is only concerned with minting new tokens as requested.

What kind of local validation would you see here?

Reply:
> @Tener
> What would you say about leveraging the `MSI_ENDPOINT` env variable for that ? For instance we can generate a uuid token and instead of using `https://azure-msi.teleport.dev`as MSI_ENDPOINT we canst the <random_uuid>msi.teleport.dev as MSI_ENDPOINT and verify if a client used proper value inside local proxy ?

We could in principle set MSI_ENDPOINT to `https://azure-msi.teleport.dev/<secure hex>`. This would be annoying in some scenarios though, for the same reason random `--port` in the `tsh proxy azure`. You need to propagate the configuration to third-party tools and if the configuration changes on each restart, this is a bad UX. But maybe we can mitigate this in similar fashion to `--port` flag, i.e. by allowing the secret to be read from env variable?

Reply:
> Right now we have several services that works with tunnel mode without local validate mode.

Sounds good.

Another reason I brought up this topic as I recall this issue #16229 an edge case for AWS, not sure this would apply to Azure though.

So say user runs `tsh app login` as AWS role A and launched `tsh proxy aws` for terraform. Terraform may use role A to assume an role B for certain operations. Obvisouly our local proxy gets requests signed for both role A and role B. We currently error out for requests signed for role B but ideally we can just route traffic for role B directly to AWS without forwarding to Teleport backend.

Reply:
We now set the `MSI_ENDPOINT` to `https://azure-msi.teleport.dev/<random hex>`, unless the user has `MSI_ENDPOINT` already set to `https://azure-msi.teleport.dev/<user chosen secret>` in which case we will use the provided secret.